Apf fix watch panic handling

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
This PR (a) introduces and uses controlflow.TryFinally to make the control flow clearer and (b) simplifying and clarifying the signalling between the goroutine that the APF handler forks in the WATCH case and the main goroutine while also adding handling of the case of both goroutines panicking. Part (b) is structured as two commits: the first removes the relaying of a panic, the second restores the relay and handles all four panic/no-panic cases.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #127048 

#### Special notes for your reviewer:
For a version without introducing and using `controlflow.TryFinally`, see #127762.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Introduce and use controlflow.TryFinally

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer committed:
Stop relay of panicked value

Instead of relaying the panicked value, log it.

Simplify the synchronization to two points in time: when the decision
of whether or not to dispatch has been made, and when the forked
goroutine is done. The first replaces the `shouldStartWatchCh` channel
with a decision-has-been-made channel. The second replaces the
panic-relay channel with a `sync.WaitGroup`.

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer commented:
The deltas shown by GitHub may be a little hard to follow, but if you look at the revised code I think that you will see that using controlflow.TryFinally makes it easier to follow (making the order of the source agree with the order of execution).

neolit123 commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -31,6 +31,7 @@ import (
 	epmetrics "k8s.io/apiserver/pkg/endpoints/metrics"
 	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/server/httplog"
+	cflow "k8s.io/apiserver/pkg/util/controlflow"
```
this could be just 
```suggestion
	"k8s.io/apiserver/pkg/util/controlflow"
```

MikeSpreitzer replied:
You're right. I have made that change.

neolit123 commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -264,24 +251,48 @@ func (h *priorityAndFairnessHandler) Handle(w http.ResponseWriter, r *http.Reque
-			watchCtx := utilflowcontrol.WithInitializationSignal(ctx, watchInitializationSignal)
-			watchReq = r.WithContext(watchCtx)
-			h.handler.ServeHTTP(w, watchReq)
+		cflow.TryFinally(func() {
```
using TryFinally instead of the defer SGTM, but i don't understand the APF logic, so leaving that to the maintainers to review.

seans3 commented:
Please have a look at PR: https://github.com/kubernetes/kubernetes/pull/127089 which is also associated with the issue.

MikeSpreitzer commented:
My recent updates are explained at https://github.com/kubernetes/kubernetes/pull/127029#issuecomment-2348219023 .

MikeSpreitzer commented:
@MikeSpreitzer

tkashem commented:
Can we address these concerns in separate PRs:
- a) the refactor only try-finally pattern, it's across multiple packages, this can go in a separate PR. I personally think it does not make the go code idiomatic, inline func with defer seems to be idiomatic and a prevalent style. However, if folks like this pattern I am fine with it (an alternate is to refactor out the standard and watch handler via a strategy, this might make the code easier to read).
- b) the bookkeeping issue warrants a separate PR in my opinion, the issue is reproduced, and the fix is backed by `stress` test result. If by any chance it introduces other flake, it can be reverted atomically. 
- c) the other changes in this PR, looks to be refactor only, if it is in a separate PR it would be easier to review 

I think these can be done in any order, thoughts?

MikeSpreitzer commented:
I just realized that I had forgotten to update the opening comment of this PR when I added the latest commits. I have done that now.

@tkashem: I think there are two major changes in this PR, as now outlined in the opening comment, and they could indeed be done in separate PRs.

MikeSpreitzer commented:
The force-push to aaa2cf8f11e simplifies the func that closes `decidedCh`: because only one goroutine invokes it, a simple boolean can be used instead of a `sync.Once`.

MikeSpreitzer committed:
Restore relaying of panic from wait goroutine

Now with hopefully simpler and clearer control flow.

Introducing stack-formatting utility that can be used
elsewhere.

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer commented:
The most recent push adds a commit that restores relaying of a panic from the wait goroutine. The code in that goroutine is not trivial, so a panic is a possibility to cover. The relaying is so that it is handled in the apiserver's normal way (by `WithPanicRecovery`). Hopefully the revised code structure is clearer.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -264,24 +249,40 @@ func (h *priorityAndFairnessHandler) Handle(w http.ResponseWriter, r *http.Reque
+			} else if waitPanicked {
+				panic("APF wait goroutine panicked: " + waitReport)
+			}
+		})
```
Yes, but I think it is more complicated than necessary. Note that `panic.go` adds more lines than all of this PR (which also removes more lines than #127029).

If we wanted, we could collect 273--284 here into a func and offer it for general use. That would be simple and concentrate that concern outside this func.

MikeSpreitzer commented:
The force-push to 3cfc7a708ad squashes out an uninteresting commit.

MikeSpreitzer commented:
The force-push to 120a5e964c2 is a rebase onto `master`.
